### PR TITLE
Remove hotplug VMI API

### PR DESCRIPTION
**Pull request by AlonaKaplan**
**What this PR does / why we need it**:
This PR removes the rest-api and virctlcommands  for hotplugging/unplugging an interfaces to/from a VMI.
Interface hotplugging/unplugging will be supported via VMs only.

It was noticed that allowing users to access the VMI (through the virt-api) is introducing too many corner cases (for example interface hotplugged to a VMI didn't get MAC from kubemacpool) and complexity in the system. Users should have access only to their user facing API, which is the VM object. Allowing the system to apply the desired spec in the VM to the backend VMI.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
Disable network interface hotplug/unplug for VMIs. It will be supported for VMs only. 
```

### CODE REVIEW

File: pkg/virt-api/rest/interfacehotplug_test.go
Comment (EdDev) [inline pkg/virt-api/rest/interfacehotplug_test.go:181]: No need for a table now.

File: pkg/virt-controller/watch/vm.go
Comment (EdDev) [inline pkg/virt-controller/watch/vm.go:2717]: The `_` is redundant.

File: pkg/virt-controller/watch/vm.go
Comment (EdDev) [inline pkg/virt-controller/watch/vm.go:2725]: Why a copy is used?

**Review by EdDev — APPROVED** [review]
Very nice, thanks!

> virtctl: remove hotplug/ugnplug from VMI
>
> hotplug/ugnplug will be allowed from VM only

Typo

File: pkg/virt-api/rest/interfacehotplug_test.go
Comment (AlonaKaplan) [reply pkg/virt-api/rest/interfacehotplug_test.go:181]: done.

File: pkg/virt-controller/watch/vm.go
Comment (AlonaKaplan) [reply pkg/virt-controller/watch/vm.go:2717]: done.

File: pkg/virt-controller/watch/vm.go
Comment (AlonaKaplan) [reply pkg/virt-controller/watch/vm.go:2725]: Since `c.VMIInterfacesPatch`compares between the changed copy and the original VMI to know if there is a need to patch the VMI.

File: pkg/virt-controller/watch/vm.go
Comment (AlonaKaplan) [reply pkg/virt-controller/watch/vm.go:2729]: I think the VMI shouldn't be changed, it represents the original VMI. Other hot updates doesn't change it as well (volume/ cpu).

File: pkg/virt-controller/watch/vm.go
Comment (AlonaKaplan) [reply pkg/virt-controller/watch/vm.go:2738]: I prefer to keep it as is since it is aligned with other patches the controller has. e.g `VMICPUsPatch`.
Changed the method no to be exported.

File: pkg/virt-controller/watch/vm.go
Comment (EdDev) [reply pkg/virt-controller/watch/vm.go]: Well, once you patch it, it changed.
If the object is not updated, decisions taken after are incorrect, e.g. status updates.

Even in this patching you assume the object was not updated by anyone else with relations to interfaces/networks.

If others have not updated it when mutating it, we do not have to repeat the same bug.
But you can reason why we need to leave the controller with an non updated object.

**Comment by EdDev** [comment]
> Very nice, thanks!
> 
> > virtctl: remove hotplug/ugnplug from VMI
> > hotplug/ugnplug will be allowed from VM only
> 
> Typo

This has not been handled.

**Comment by maiqueb** [comment]
I don't think we can claim that the VM is only user facing API for a workload on KubeVirt.

> Users should have access only to their user facing API, which is the VM object.

I think this is not correct. 

Are we sure we don't users using this ?
@chinglinwen does this affect you in any way ?

**Comment by maiqueb** [comment]
Also, please be more specific.
> allowing users to access the VMI (through the virt-api) is introducing too many corner cases

doesn't help understand the reasons behind breaking the API.

A list of corner cases would.

**Comment by EdDev** [comment]
> I don't think we can claim that the VM is only user facing API for a workload on KubeVirt.

I think we can.

> > Users should have access only to their user facing API, which is the VM object.
> 
> I think this is not correct.
> 
> Are we sure we don't have upstream users using/relying on this ? Wouldn't it make sense to send an email to the mailing list giving a heads-up of your intentions ?
> 
> @chinglinwen does this affect you in any way ?

This API has not yet been released under a formal version.
We can and we should do this before the release is done to avoid deprecation notices.

In addition, the hotplug and unplug API/s are in tech-preview and this is how I suggest to treat it.
Equivalent to Kuberenetes alpha version if this helps give it a context.

The lesson learned from the period this API was in, is proof enough that we should fix the mistake we did sooner than later.

File: pkg/virt-controller/watch/vm.go
Comment (EdDev) [reply pkg/virt-controller/watch/vm.go]: I will follow up on this and send a PR to fix it, it is not worth blocking on it.

**Review by EdDev — APPROVED** [review]
Thank you!

**Comment by chinglinwen** [comment]
> Users should have access only to their user facing API, which is the VM object.

I have an example that we depends on VM's pods `k6t` bridge device( to flood packet to do the packet mirror), it's internal, observable behavior will be depends on somehow ;).  

> @chinglinwen does this affect you in any way ?

@maiqueb  for this particular PR, we don't using hot plug yet, we may using it later, so it currently doesn't affect us
I notice this change remove temporary hot plug feature, I'm not sure we will need it or not ( and I've work on something else currently, so I've lost some detail in my minds ).

**Comment by EdDev** [comment]
> I have an example that we depends on VM's pods `k6t` bridge device( to flood packet to do the packet mirror), it's internal, observable behavior will be depends on somehow ;).

The existing code of hotplug has already broken the assumptions taken on the internal pod networking.
The names of the interfaces used in the pod have changed from ordinal to hashed.
But I guess this is the risk of any downstream that tweaks and depends on internals.

**Comment by AlonaKaplan** [comment]
> @maiqueb for this particular PR, we don't using hot plug yet, we may using it later, so it currently doesn't affect us I notice this change remove temporary hot plug feature, I'm not sure we will need it or not ( and I've work on something else currently, so I've lost some detail in my minds ).

The PR is not removing the hotplug feature. It is still available via VMs. It just removes the option to hotplug directly to a VMI.

**Comment by chinglinwen** [comment]
> > I have an example that we depends on VM's pods `k6t` bridge device( to flood packet to do the packet mirror), it's internal, observable behavior will be depends on somehow ;).
> 
> The existing code of hotplug has already broken the assumptions taken on the internal pod networking.
> The names of the interfaces used in the pod have changed from ordinal to hashed.
> But I guess this is the risk of any downstream that tweaks and depends on internals.

Got it, thanks for the note.

**Comment by rmohr** [comment]
> I don't think we can claim that the VM is only user facing API for a workload on KubeVirt.

>> Users should have access only to their user facing API, which is the VM object.

> I think this is not correct.

I can only second @maiqueb. VMI, VM, VMIRS, VM Pools are all user-facing APIs. That the VMI is not user-facing is not correct. I would kindly ask to bring this to the mailing list and ask for broader approver commitment. Maybe they agree to start considering the VMI as non-userfacing in the future, but as it is right now that is not a correct assumption.

**Pull request merged**
